FOCUS #182: Usage Quantity (#186)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/dimensions/quantityinusageunit.md

# Quantity In Usage Unit

> Quantity In Usage Unit is the amount of resources or services consumed. The Quantity In Usage Unit is measured in units specified by the Usage Unit. Quantity In Usage Unit may differ from a Quantity In Pricing Unit when providers use a different unit of measure for calculating cost. Quantity In Usage Unit is often used when analyzing environments to gain understanding of usage or consumption of a given product/service in an environment.

Reviewer: Can we generalize to be more inclusive?
```suggestion
Quantity In Usage Unit is the amount of a resource or service that was used or purchased. The Quantity In Usage Unit is measured in units specified by the Usage Unit. Quantity In Usage Unit may differ from a Quantity In Pricing Unit when providers use a different unit of measure for calculating cost. Quantity In Usage Unit is often used when analyzing environments to gain understanding of usage or consumption of a given product/service in an environment.
```

Quantity In Usage Unit MUST exist in billing data. The column MUST be a numeric value and MUST NOT contain nulls.

## Column ID

> quantity_in_usage_unit

Reviewer: ```suggestion
QuantityInUsageUnit
```

> # Quantity In Usage Unit

Reviewer: As we've seen columns mature and close on a common prefixed grouping pattern, I would really like to see this and `QuantityInPricingUnit` be called `UsageQuantity` and `PricingQuantity` so they sit next to their respective unit columns in a list. These will be very out-of-place if not sitting next to the column that explains what the number means.

Quantity In Usage Unit is the number of units of a resource or service that was used or consumed. The Quantity In Usage Unit is measured in units specified by the Usage Unit. Quantity In Usage Unit is used in conjunction with Usage Unit to gain understanding of usage or consumption of a given product/service in an environment.

Quantity In Usage Unit MUST exist in billing data. The column MUST be a numeric value and MUST NOT contain nulls.

> The Quantity In Usage Unit for a particular line item may differ from Quantity In Pricing Unit when providers use different quantity, quantity increments, or columns to determine cost. The values in UsageUnit and QuantityInUsageUnit are often listed at a finer granularity or over a different time interval than the PricingUnit and QuantityInPricingUnit. While QuantityInUsageUnit may appear to relate to pricing and cost, QuantityInUsageUnit is focused on resource or service consumption. It is essential not to confuse QuantityInUsageUnit with QuantityInPricingUnit which is the basis for determining cost. The QuantityInUsageUnit column MUST not be used as the basis for determining values related to any pricing or cost metric.

Reviewer 2: "MUST NOT"

Author: Corrected, thanks!

> Quantity In Usage Unit MUST exist in billing data. The column MUST be a numeric value and when the [ChargeType](#chargetyep) is "Usage" Quantity In Usage Unit SHOULD NOT contain nulls. Negative values are only permissable when providers are adjusting for usage corrections.

Reviewer 3: " Quantity In Usage Unit SHOULD NOT contain nulls" implies that it *may* contain null, however column is declared as not allowing nulls

Reviewer: This column has to allow nulls for taxes and some adjustments. But I do agree this feels too soft. If we want this filled out, we should be explicit. Suggesting an update to be more complete and align with others.
```suggestion
QuantityInUsageUnit MUST exist in billing data. The column MUST be a numeric value. When [ChargeType](#chargetype) is "Usage" or "Purchase", QuantityInUsageUnit MUST NOT be null or empty. When ChargeType is "Adjustment" and the adjustment applies to specific charges that had a common UsageUnit, QuantityInUsageUnit MUST NOT be null. Negative values are only permissable when providers are adjusting for usage corrections.
```

## Column ID

QuantityInUsageUnit

## Display name

Quantity In Usage Unit

## Description

The number of units of a resource or service that was used or consumed based on the [UsageUnit](#usageunit)

## Content constraints

| Constraint      | Value        |
|-----------------|--------------|
| Column required | True         |
| Data type       | Numeric      |

> | Allows nulls    | False        |

Reviewer: ```suggestion
| Allows nulls    | True         |
```
